go/vt/vtgate/plugin_mysql_server.go

```diff
@@ -216,18 +225,31 @@ func newMysqlUnixSocket(address string, authServer mysql.AuthServer, handler mys
 func init() {
 	servenv.OnRun(initMySQLProtocol)
 
-	servenv.OnTerm(func() {
+	servenv.OnTermSync(func() {
+		wg := &sync.WaitGroup{}
 		if mysqlListener != nil {
-			mysqlListener.Close()
-			mysqlListener = nil
+			drainListener(mysqlListener, wg)
 		}
 		if mysqlUnixListener != nil {
-			mysqlUnixListener.Close()
-			mysqlUnixListener = nil
+			drainListener(mysqlUnixListener, wg)
 		}
+		log.Info("Waiting for mysql connections to drain...")
+		wg.Wait()
+		log.Info("All mysql connections drained successfully!")
 	})
 }
 
+func drainListener(l *mysql.Listener, wg *sync.WaitGroup) {
+	wg.Add(1)
+	go func() {
+		l.Shutdown()
```

Comment:
This is not required. You can actually delete all the logic involved with Shutdown. Bascially, l.Close() already does all this, which is to stop listening.